### ISSUE TITLE
Add `arc_read()` function

### ISSUE THREAD

**Issue by elipousson** [R-ArcGIS/arcgislayers#108]
I would find it really convenient to have a single function, e.g. `arc_read()`, that I could use to read in data from a ArcGIS service. Here is a reprex showing my existing workflow and an alternate workflow with this function:

``` r
library(arcgislayers)
library(rlang) # for %||%

url <- "https://geodata.baltimorecity.gov/egis/rest/services/CitiMap/DOT_Layers/MapServer/5"

# The way I usually use arcgislayers
data <- url |>
  arc_open() |>
  arc_select()

# The additional convenience function for reading layers
arc_read <- function(url,
                     fields = NULL,
                     where = NULL,
                     crs = NULL,
                     geometry = TRUE,
                     filter_geom = NULL,
                     ...,
                     .name_repair = NULL,
                     token = Sys.getenv("ARCGIS_TOKEN")) {
  arc_obj <- arc_open(url = url, token = token)

  arc_obj <- arc_select(
    x = arc_obj,
    fields = fields,
    where = where,
    crs = crs %||% sf::st_crs(arc_obj),
    geometry = geometry,
    filter_geom = filter_geom,
    token = token,
    ...
  )

  if (!is.null(.name_repair)) {
    if (is.function(.name_repair)) {
      nm <- .name_repair(names(arc_obj))
    }

    arc_obj <- rlang::set_names(arc_obj, nm)
  }

  arc_obj
}

url <- "https://geodata.baltimorecity.gov/egis/rest/services/CitiMap/DOT_Layers/MapServer/5"

data <- arc_read(url)
```

<sup>Created on 2023-11-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

This reprex also shows the potential of adding a .name_repair parameter. If you'd be OK with picking up a `{vctrs}` dependency you could use `vctrs::vec_as_names()` or re-implement comparable flexibility.

**Comment by elipousson** [comment]
I don't think `arc_open() |> arc_select()` is too much hassle but I think a single function would be useful—especially when so many vector layers are fairly small and read into memory quickly without any issues.

I'd also note that most of the common read functions don't *require* you to read everything into memory:

- `sf::read_sf()` supplies a `wkt_filter` parameter that is similar to `filter_geom` (and you can use the `query` parameter to select specific columns or attributes or filter rows)
- `readr::read_delim()` has `col_select` (comparable to `fields`) and `skip`
- `googlesheets4::read_sheet()` also has `range` and `skip`
- `arrow::read_parquet()` works with URLs as well as any Arrow input stream and also has `col_select`

`arrow::open_dataset()` is designed for multi-file datasets which is why the limitation makes sense there. You could have `arc_read()` kick back an error if the feature count from the initial metadata exceeds some default threshold.

**Comment by JosiahParry** [comment]
Let's go ahead and make a pr. Why not? be sure to use `rlang::names_inform_repair()` if names are changed. Let's set `n_max = 10000` as a safe maximum (I've seen some insanely detailed polylines recently where 300 take longer to read in 5,000 polygons). After using `arc_open()` its probably best to use `obj_is_layer()` in case somehow we read in a map or feature server.

**Comment by elipousson** [comment]
Nice. Thanks! Would you be interested in me including the col_select argument as an alias for fields? Internally inconsistent but maybe nice for familiarity based on those other read functions.

**Comment by elipousson** [comment]
That is almost exactly how I have it started. Stayed up late getting it working (getting col_names to parity with the standard tidyverse functionality was a bit tricky) but the PR may not be until next weekend. I think it could also easily support `arc_raster` (that is how I wrote it so far)

**Comment by JosiahParry** [comment]
Closed by #118
